rescript/get_gtdb.py

```diff
@@ -0,0 +1,138 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2019-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import os
+import tempfile
+import tarfile
+import warnings
+
+import qiime2
+from urllib.request import urlretrieve
+from collections import defaultdict
+from urllib.error import HTTPError
+from rescript.get_data import _gzip_decompress
+
+# Different versions may have different file names for archaea and
+# bacteria. for example 'ar53' and 'bac120' mean that the GTDB phylogeny
+# is based on 53 and 120 concatenated proteins (cp), respectively.
+# If this changes we can set up a conditional statemnt below.
+VERSION_MAP_DICT = {'207': {'Archaea': 'ar53', 'Bacteria': 'bac120'},
+                    '202': {'Archaea': 'ar122', 'Bacteria': 'bac120'}}
+
+
+def get_gtdb_data(ctx, version='207', domain='Both'):
+
+    ver_dom_dict = defaultdict(lambda: defaultdict(dict))
+
+    # Subset dict if needed, but keep same structure
+    # Although we can run the following merge actions on a list of one
+    # i.e. 'Archaea', we do not want to confuse anyone when looking
+    # at provenance, by running a merge command for no reason.
+    if domain == 'Both':
+        ver_dom_dict[version] = VERSION_MAP_DICT[version]
+    else:
+        ver_dom_dict[version][domain] = VERSION_MAP_DICT[version][domain]
+
+    queries = _assemble_queries(ver_dom_dict)
+    tax_q, seqs_q = _retrieve_data_from_gtdb(queries)
+
+    if domain == 'Both':
+        merge_gtdb_seqs = ctx.get_action('feature_table', 'merge_seqs')
+        merge_gtdb_taxonomy = ctx.get_action('feature_table', 'merge_taxa')
+        print('\n  Merging taxonomy data...')
+        merged_gtdb_tax, = merge_gtdb_taxonomy(data=tax_q)
+        print('\n  Merging sequence data...')
+        merged_gtdb_seqs, = merge_gtdb_seqs(data=seqs_q)
+    else:
+        merged_gtdb_tax = tax_q[0]
+        merged_gtdb_seqs = seqs_q[0]
+
+    print('\n Saving files...\n')
+    return merged_gtdb_tax, merged_gtdb_seqs
+
+
+def _assemble_queries(ver_dom_dict):
+    queries = defaultdict(lambda: list())
+
+    base_seq_url = ('https://data.gtdb.ecogenomic.org/releases/release%s/'
+                    '%s.0/genomic_files_reps/%s_ssu_reps_r%s.tar.gz')
+    base_tax_url = ('https://data.gtdb.ecogenomic.org/releases/release%s/'
+                    '%s.0/%s_taxonomy_r%s.tsv.gz')
+
+    for version, dcp in ver_dom_dict.items():
+        for dom, cp in dcp.items():
+
+            queries['Taxonomy'].append(
+                (dom,
+                 base_tax_url % (version, version, cp, version),
```

Comment:
Is there any reason to use this way of string formatting and the `format` method everywhere else below? Maybe you could use `format` here too, for consistency?

Reply:
Nice catch! No reason at all... just a momentary lapse... :-)